### PR TITLE
ftrace: improve binary parser speed by optimizing I/O

### PR DESCRIPTION
**Pull request by fabrizioiannetti**
* mmap the whole file for faster parsing
* re-use the same mapping to read events

on my (oldish) laptop with ubuntu 24.04 the file is parsed an order of magnitude faster

on a recent windows laptop I measured two orders of magnitude faster

### CODE REVIEW

**Comment by MatthewKhouzam** [comment]
This is still WIP. I know your approach will work, we didn't have time to get it done. (We start by coding defensively, then optimize) This will be great, please remove the printlns though ;)

**Comment by fabrizioiannetti** [comment]
> This is still WIP. I know your approach will work, we didn't have time to get it done. (We start by coding defensively, then optimize) This will be great, please remove the printlns though ;)

missed that one :-), removed the println.

**Comment by fabrizioiannetti** [comment]
Hi, is something missing in the PR for review?

**Review by MatthewKhouzam — DISMISSED** [review]
Missed the update, sorry!

File: tracetypes/org.eclipse.tracecompass.incubator.ftrace.core/src/org/eclipse/tracecompass/incubator/internal/ftrace/core/trace/BinaryFTrace.java [hunk 1]
```diff
@@ -115,6 +115,14 @@ public void initTrace(IResource resource, String path, Class<? extends ITmfEvent
         setFile(file);
     }
 
+    @Override
+    public synchronized void dispose() {
+        super.dispose();
+        // release (indirect) references to mem-mapped file buffers so that tracecompass
+        // can garbage-collect them and unlock the file (e.g. if the user wants to delete it)
+        fStrategy = null;
```
Comment (arfio) [inline tracetypes/org.eclipse.tracecompass.incubator.ftrace.core/src/org/eclipse/tracecompass/incubator/internal/ftrace/core/trace/BinaryFTrace.java:123]: There should be a dispose method in the IBinaryFTraceStrategy that does this for the BinaryFTraceV6Strategy.

**Review by arfio — CHANGES_REQUESTED** [review]
This change looks great, there is just one more change needed and I then I can approve it 😄

File: tracetypes/org.eclipse.tracecompass.incubator.ftrace.core/src/org/eclipse/tracecompass/incubator/internal/ftrace/core/trace/BinaryFTrace.java (hunk 1 above)
Comment (fabrizioiannetti) [reply tracetypes/org.eclipse.tracecompass.incubator.ftrace.core/src/org/eclipse/tracecompass/incubator/internal/ftrace/core/trace/BinaryFTrace.java]: Thanks for the review, done: I moved the clearing of the reference to the strategy class.
I did not go deeper in the structure, let me know if it's good like this.

Comment (fabrizioiannetti) [reply tracetypes/org.eclipse.tracecompass.incubator.ftrace.core/src/org/eclipse/tracecompass/incubator/internal/ftrace/core/trace/BinaryFTrace.java]: seems I broke something with that, need to look into it :-)

Comment (arfio) [reply tracetypes/org.eclipse.tracecompass.incubator.ftrace.core/src/org/eclipse/tracecompass/incubator/internal/ftrace/core/trace/BinaryFTrace.java]: The change is perfect, it just needs to pass the test now :)

Comment (arfio) [reply tracetypes/org.eclipse.tracecompass.incubator.ftrace.core/src/org/eclipse/tracecompass/incubator/internal/ftrace/core/trace/BinaryFTrace.java]: Seems like it needs a null check before calling dispose().

Comment (fabrizioiannetti) [reply tracetypes/org.eclipse.tracecompass.incubator.ftrace.core/src/org/eclipse/tracecompass/incubator/internal/ftrace/core/trace/BinaryFTrace.java]: Thanks for the hint! Took me a while to find the right logs

**Review by arfio — APPROVED** [review]

**Review by MatthewKhouzam — APPROVED** [review]

**Comment by MatthewKhouzam** [comment]
@fabrizioiannetti thank you so much for the contribution!

**Pull request merged**